tsdb/head.go

```diff
@@ -1833,23 +1898,80 @@ func (s *memSeries) iterator(id int, it chunkenc.Iterator) chunkenc.Iterator {
 		return chunkenc.NewNopIterator()
 	}
 
+	ix := id - s.firstChunkID
+
+	numSamples := c.chunk.NumSamples()
+	stopAfter := numSamples
+
+	if isoState != nil {
+		totalSamples := 0    // Total samples in this series.
+		previousSamples := 0 // Samples before this chunk.
+
+		for j, d := range s.chunks {
+			totalSamples += d.chunk.NumSamples()
+			if j < ix {
+				previousSamples += d.chunk.NumSamples()
+			}
+		}
+
+		// Removing the extra transactionIDs that are relevant for samples that
+		// come after this chunk, from the total transactionIDs.
+		writeIDsToConsider := s.txs.txIDCount - (totalSamples - (previousSamples + numSamples))
+
+		// Iterate over the writeIDs, find the first one that the isolation state says not
+		// to return.
+		it := s.txs.iterator()
+		for index := 0; index < writeIDsToConsider; index++ {
+			writeID := it.At()
+			if writeID <= isoState.maxWriteID { // Easy check first.
+				if _, ok := isoState.incompleteWrites[writeID]; !ok {
+					it.Next()
+					continue
+				}
+			}
+			stopAfter = numSamples - (writeIDsToConsider - index)
+			if stopAfter < 0 {
+				stopAfter = 0 // Stopped in a previous chunk.
+			}
+			break
+		}
+	}
+
 	if id-s.firstChunkID < len(s.chunks)-1 {
-		return c.chunk.Iterator(it)
+		// TODO(beorn7): Reviewers, please double-check this section extra carefully.
+		if msIter, ok := it.(*memSafeIterator); ok {
+			msIter.Iterator = c.chunk.Iterator(msIter.Iterator)
+			msIter.i = -1
+			msIter.total = numSamples
+			msIter.stopAfter = stopAfter
+			msIter.bufferedSamples = 0
```

Comment:
No buf needed?

Reply:
I assume, yes.

Reply:
No buffered samples at that moment, if I understand that correctly.

Reply:
Yeah, we only need to do this for the last chunk as it's the only one that has buffered samples.

Reply:
Thanks everyone.